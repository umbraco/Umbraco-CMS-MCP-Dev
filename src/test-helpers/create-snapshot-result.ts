import { BLANK_UUID } from "@/constants/constants.js";
import { CallToolResult } from "@modelcontextprotocol/sdk/types.js";

export function createSnapshotResult(result: any, idToReplace?: string) {
  if (!result?.content) {
    return result;
  }

  function normalizeItem(i: any) {
    const item = { ...i, id: BLANK_UUID };
    if (item.parent) {
      item.parent = { ...item.parent, id: BLANK_UUID };
    }
    if (item.createDate) {
      item.createDate = "NORMALIZED_DATE";
    }
    if (item.publishDate) {
      item.publishDate = "NORMALIZED_DATE";
    }
    if (item.updateDate) {
      item.updateDate = "NORMALIZED_DATE";
    }
    return item;
  }

  return {
    ...result,
    content: result.content.map((item: any) => {
      if (item.type === "text") {
        if (idToReplace) {
          // For single item responses
          let text = item.text.replace(idToReplace, BLANK_UUID);
          try {
            const parsed = JSON.parse(text);
            if (parsed.createDate) {
              parsed.createDate = "NORMALIZED_DATE";
            }
            if (parsed.availableUntil) {
              parsed.availableUntil = "NORMALIZED_DATE";
            }
            if (parsed.publishDate) {
              parsed.publishDate = "NORMALIZED_DATE";
            }
            if (parsed.updateDate) {
              parsed.updateDate = "NORMALIZED_DATE";
            }
            if (parsed.variants && Array.isArray(parsed.variants)) {
              parsed.variants = parsed.variants.map((variant: any) => {
                if (variant.createDate) variant.createDate = "NORMALIZED_DATE";
                if (variant.publishDate)
                  variant.publishDate = "NORMALIZED_DATE";
                if (variant.updateDate) variant.updateDate = "NORMALIZED_DATE";
                return variant;
              });
            }
            text = JSON.stringify(parsed);
          } catch {}
          return {
            ...item,
            text,
          };
        } else {
          // For list responses
          const parsed = JSON.parse(item.text);
          if (Array.isArray(parsed)) {
            // Handle ancestors API response
            return {
              ...item,
              text: JSON.stringify(parsed.map(normalizeItem)),
            };
          }
          // Handle other list responses
          if (parsed.items) {
            parsed.items = parsed.items.map(normalizeItem);
          }
          if (parsed.variants && Array.isArray(parsed.variants)) {
            parsed.variants = parsed.variants.map((variant: any) => {
              if (variant.createDate) variant.createDate = "NORMALIZED_DATE";
              if (variant.publishDate) variant.publishDate = "NORMALIZED_DATE";
              if (variant.updateDate) variant.updateDate = "NORMALIZED_DATE";
              return variant;
            });
          }
          return {
            ...item,
            text: JSON.stringify(parsed),
          };
        }
      }
      return item;
    }),
  };
}

export function normalizeErrorResponse(result: CallToolResult): CallToolResult {
  if (
    Array.isArray(result.content) &&
    result.content[0]?.text &&
    typeof result.content[0].text === "string"
  ) {
    // Replace any traceId in the text with a normalized version
    result.content[0].text = result.content[0].text.replace(
      /00-[0-9a-f]{32}-[0-9a-f]{16}-00/g,
      "normalized-trace-id"
    );
  }
  return result;
}
