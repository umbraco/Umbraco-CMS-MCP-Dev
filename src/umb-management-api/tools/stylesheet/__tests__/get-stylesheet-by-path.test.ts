import GetStylesheetByPathTool from "../get/get-stylesheet-by-path.js";
import { StylesheetHelper } from "./helpers/stylesheet-helper.js";
import { StylesheetBuilder } from "./helpers/stylesheet-builder.js";
import { createSnapshotResult } from "@/test-helpers/create-snapshot-result.js";
import { jest } from "@jest/globals";

const TEST_STYLESHEET_NAME = "_TestGetStylesheet.css";
const TEST_CONTENT = "/* Test get stylesheet */\nbody { background: white; }\n.container { padding: 20px; }";
const NON_EXISTENT_STYLESHEET_PATH = "/_NonExistentStylesheet.css";

describe("get-stylesheet-by-path", () => {
  let originalConsoleError: typeof console.error;

  beforeEach(() => {
    originalConsoleError = console.error;
    console.error = jest.fn();
  });

  afterEach(async () => {
    await StylesheetHelper.cleanup(TEST_STYLESHEET_NAME);
    console.error = originalConsoleError;
  });

  it("should get a stylesheet by path", async () => {
    // Arrange - Create a stylesheet first using builder
    const stylesheetBuilder = new StylesheetBuilder()
      .withName(TEST_STYLESHEET_NAME)
      .withContent(TEST_CONTENT);
    
    await stylesheetBuilder.create();
    const stylesheetPath = stylesheetBuilder.getPath();

    // Act
    const result = await GetStylesheetByPathTool().handler({ path: stylesheetPath }, { signal: new AbortController().signal });

    // Assert
    const normalizedResult = createSnapshotResult(result);
    expect(normalizedResult).toMatchSnapshot();

    // Verify content is preserved with CSS formatting
    const parsedResult = JSON.parse(result.content[0].text as string);
    expect(parsedResult.content).toContain("background: white;");
    expect(parsedResult.content).toContain("padding: 20px;");
    expect(parsedResult.name).toBe(TEST_STYLESHEET_NAME);
    expect(parsedResult.path).toBe(stylesheetPath);
  });

  it("should handle non-existent stylesheet", async () => {
    // Act
    const result = await GetStylesheetByPathTool().handler({ path: NON_EXISTENT_STYLESHEET_PATH }, { signal: new AbortController().signal });

    // Assert
    expect(result).toMatchSnapshot();
  });
});