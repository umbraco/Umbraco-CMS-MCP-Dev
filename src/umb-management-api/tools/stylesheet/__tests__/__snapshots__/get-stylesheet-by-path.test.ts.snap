// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-stylesheet-by-path should get a stylesheet by path 1`] = `
{
  "content": [
    {
      "text": "{"content":"/* Test get stylesheet */\\nbody { background: white; }\\n.container { padding: 20px; }","name":"_TestGetStylesheet.css","parent":null,"path":"/_TestGetStylesheet.css"}",
      "type": "text",
    },
  ],
}
`;

exports[`get-stylesheet-by-path should handle non-existent stylesheet 1`] = `
{
  "content": [
    {
      "text": "Error using get-stylesheet-by-path:
{
  "message": "Request failed with status code 404",
  "response": {
    "type": "Error",
    "title": "Stylesheet not found",
    "status": 404,
    "detail": "The stylesheet was not found.",
    "operationStatus": "NotFound"
  }
}",
      "type": "text",
    },
  ],
}
`;
